test/distributed/test_hooks.py

```diff
@@ -138,6 +141,65 @@ def check_op(idx, coll_name):
         check_op(0, "ALLGATHER")
         check_op(1, "ALLREDUCE")
 
+    def _test_failure(self):
+        # it's ok to access them directly since there's a single bg thread poking at them.
+        starts = []
+        ends = []
+        cv = threading.Condition()
+
+        def coll_start(status):
+            starts.append(status)
+            print(f"col_start {len(starts)} rank{self.rank}")
+
+        def coll_end(status):
+            ends.append(status)
+            print(f"col_end {len(ends)} rank{self.rank}")
+            if len(ends) == 2:
+                with cv:
+                    cv.notify()
+
+        dhooks.register_collective_start_hook(coll_start)
+        dhooks.register_collective_end_hook(coll_end)
+
+        tensor = torch.ones([2, 3]).cuda(self.rank) * self.rank
+        tensor_list = [torch.empty_like(tensor) for _ in range(self.world_size)]
+
+        dist.all_gather(tensor_list, tensor)
+
+        tensor2 = torch.ones([2, 3]).cuda(self.rank) * self.rank
+        if self.rank == 2:
+            time.sleep(10)
+        dist.all_reduce(tensor2)
+
+        with cv:
+            cv.wait(20)
+
+        default_pg_name = dist.group.WORLD.group_name
+        self.assertEqual(2, len(starts))
+        self.assertEqual(2, len(ends))
+
+        def check_op(idx, coll_name, end_fail=False):
+            self.assertEqual(default_pg_name, starts[idx].pg_name)
+            self.assertEqual(self.backend_name, starts[idx].backend)
+            self.assertGreaterEqual(starts[idx].sequence_number, 0)
+            self.assertGreaterEqual(starts[idx].timestamp, 0)
+            self.assertEqual(coll_name, starts[idx].operation)
+            self.assertEqual(starts[idx].error_message, None)
+            if end_fail:
+                self.assertIsNotNone(ends[idx].error_message, None)
```

Comment:
perhaps assert some known string is present in the error msg too?
(and that might also answer my question: what error are we expecting in this example?)

Reply:
good point, hopefully it will be reliable enough during testing.